Implement lazy decoding of DefPathTable during incremental compilation

Aaron1011 opened the pull request:
PR https://github.com/rust-lang/rust/pull/75813 implemented lazy decoding of the `DefPathTable` from crate metadata. However, it requires decoding the entire `DefPathTable` when incremental compilation is active, so that we can map a decoded `DefPathHash` to a `DefId` from an arbitrary crate.

This PR adds support for lazy decoding of dependency `DefPathTable`s when incremental compilation si active.

When we load the incremental cache and dep
graph, we need the ability to map a `DefPathHash` to a `DefId` in the
current compilation session (if the corresponding definition still
exists).

This is accomplished by storing the old `DefId` (that is, the `DefId`
from the previous compilation session) for each `DefPathHash` we need to
remap. Since a `DefPathHash` includes the owning crate, the old crate is
guaranteed to be the right one (if the definition still exists). We then
use the old `DefIndex` as an initial guess, which we validate by
comparing the expected and actual `DefPathHash`es. In most cases,
foreign crates will be completely unchanged, which means that we our
guess will be correct. If our guess is wrong, we fall back to decoding
the entire `DefPathTable` for the foreign crate. This still represents
an improvement over the status quo, since we can skip decoding the
entire `DefPathTable` for other crates (where all of our guesses were
correct).

rust-highfive commented:
r? @petrochenkov

(rust_highfive has picked a reviewer for you, use r? to override)

Aaron1011 commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion

bors commented:
:hourglass: Trying commit c169390670b5400b83c79abc8662ffd76eaa321a with merge efc44d114215b797e4e501a021a02c4ab500aea0...
<!-- homu: {"type":"TryBuildStarted","head_sha":"c169390670b5400b83c79abc8662ffd76eaa321a","merge_sha":"efc44d114215b797e4e501a021a02c4ab500aea0"} -->

Aaron1011 commented:
@bors try-

Aaron1011 commented:
@bors untry

rust-log-analyzer commented:
The job `dist-x86_64-linux` of your PR [failed](https://github.com/rust-lang-ci/rust/runs/930474174) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://github.com/rust-lang-ci/rust/commit/efc44d114215b797e4e501a021a02c4ab500aea0/checks/930474174/logs), [raw log](https://github.com/rust-lang-ci/rust/commit/efc44d114215b797e4e501a021a02c4ab500aea0/checks/930474174/logs)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
   Compiling tempfile v3.1.0
   Compiling serde_derive v1.0.106
   Compiling serde_json v1.0.40
   Compiling rustdoc v0.0.0 (/checkout/src/librustdoc)
error: unused import: `rustc_middle::middle::cstore::CrateStore`
   |
   |
17 | use rustc_middle::middle::cstore::CrateStore;
   |
   |
   = note: `-D unused-imports` implied by `-D warnings`

error[E0599]: no method named `def_path_table` found for reference `&rustc_metadata::creader::CStore` in the current scope
    |
    |
123 |                 self.enter_resolver(|r| r.cstore().def_path_table(crate_num).next_id())
    |                                                    ^^^^^^^^^^^^^^ method not found in `&rustc_metadata::creader::CStore`
error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0599`.


Did not run successfully: exit code: 1
"/checkout/obj/build/x86_64-unknown-linux-gnu/stage1/bin/rustc" "--crate-name" "rustdoc" "--edition=2018" "src/librustdoc/lib.rs" "--error-format=json" "--json=diagnostic-rendered-ansi" "--crate-type" "lib" "--emit=dep-info,metadata,link" "-C" "opt-level=3" "-Cembed-bitcode=no" "-C" "debuginfo=0" "-C" "metadata=c4e25bbdb2cbfbd7" "-C" "extra-filename=-c4e25bbdb2cbfbd7" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/deps" "--target" "x86_64-unknown-linux-gnu" "-C" "linker=clang" "-L" "dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/deps" "-L" "dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/release/deps" "--extern" "itertools=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/deps/libitertools-dd82d5121748193a.rmeta" "--extern" "minifier=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/deps/libminifier-966c1716170dd1ef.rmeta" "--extern" "pulldown_cmark=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/deps/libpulldown_cmark-56267caff99f6f9f.rmeta" "--extern" "rayon=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/deps/librustc_rayon-66f8d44f6c3978ea.rmeta" "--extern" "serde=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/deps/libserde-13a0e25aeb4de8b3.rmeta" "--extern" "serde_json=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/deps/libserde_json-fb1ef14d4c222746.rmeta" "--extern" "tempfile=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/deps/libtempfile-12144885b7f5530c.rmeta" "-Zmacro-backtrace" "-Clink-args=-Wl,-rpath,$ORIGIN/../lib" "-Zbinary-dep-depinfo" "-Wrust_2018_idioms" "-Wunused_lifetimes" "-Dwarnings" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1" "--remap-path-prefix" "/checkout=/rustc/efc44d114215b797e4e501a021a02c4ab500aea0"
error: could not compile `rustdoc`.

To learn more, run the command again with --verbose.

---
== clock drift check ==
  local time: Fri Jul 31 04:06:42 UTC 2020
  network time: Fri, 31 Jul 2020 04:06:42 GMT
== end clock drift check ==
##[error]Process completed with exit code 1.
Terminate orphan process: pid (4713) (python)
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@rust-lang/infra`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

Aaron1011 commented:
@bors try

bors commented:
:hourglass: Trying commit 983d06c6ac41a906cfb178e46a21a2fe5e0a561a with merge db610e4857f4482661cb2f4a30150f12ec9a6cca...
<!-- homu: {"type":"TryBuildStarted","head_sha":"983d06c6ac41a906cfb178e46a21a2fe5e0a561a","merge_sha":"db610e4857f4482661cb2f4a30150f12ec9a6cca"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/930607063), [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=35707)
Build commit: db610e4857f4482661cb2f4a30150f12ec9a6cca (`db610e4857f4482661cb2f4a30150f12ec9a6cca`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=35707","checks-actions":"https://github.com/rust-lang-ci/rust/runs/930607063"},"merge_sha":"db610e4857f4482661cb2f4a30150f12ec9a6cca"} -->

rust-timer commented:
Queued db610e4857f4482661cb2f4a30150f12ec9a6cca with parent c058a8b8dc5dea0ed9b33e14da9e317e2749fcd7, future [comparison URL](https://perf.rust-lang.org/compare.html?start=c058a8b8dc5dea0ed9b33e14da9e317e2749fcd7&end=db610e4857f4482661cb2f4a30150f12ec9a6cca).

rust-timer commented:
Finished benchmarking try commit (db610e4857f4482661cb2f4a30150f12ec9a6cca): [comparison url](https://perf.rust-lang.org/compare.html?start=c058a8b8dc5dea0ed9b33e14da9e317e2749fcd7&end=db610e4857f4482661cb2f4a30150f12ec9a6cca).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. Please note that if the perf results are neutral, you should likely undo the rollup=never given below by specifying `rollup-` to bors.

Importantly, though, if the results of this run are non-neutral **do not** roll this PR up -- it will mask other regressions or improvements in the roll up.

@bors rollup=never

petrochenkov commented:
Looks like a regression for larger crates, but win for tiny crates.
(Besides the `*-doc` anomaly.)

Aaron1011 commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion

bors commented:
:hourglass: Trying commit c581132756dd09ad72ff2e694d63441d70f2e201 with merge 6616659d34a250401009ff4286292cf29487552b...
<!-- homu: {"type":"TryBuildStarted","head_sha":"c581132756dd09ad72ff2e694d63441d70f2e201","merge_sha":"6616659d34a250401009ff4286292cf29487552b"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/931821104), [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=35712)
Build commit: 6616659d34a250401009ff4286292cf29487552b (`6616659d34a250401009ff4286292cf29487552b`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=35712","checks-actions":"https://github.com/rust-lang-ci/rust/runs/931821104"},"merge_sha":"6616659d34a250401009ff4286292cf29487552b"} -->

rust-timer commented:
Queued 6616659d34a250401009ff4286292cf29487552b with parent 3a92b9987abd01c4b7e59c870e85beb9dd4d4aa2, future [comparison URL](https://perf.rust-lang.org/compare.html?start=3a92b9987abd01c4b7e59c870e85beb9dd4d4aa2&end=6616659d34a250401009ff4286292cf29487552b).

rust-timer commented:
Finished benchmarking try commit (6616659d34a250401009ff4286292cf29487552b): [comparison url](https://perf.rust-lang.org/compare.html?start=3a92b9987abd01c4b7e59c870e85beb9dd4d4aa2&end=6616659d34a250401009ff4286292cf29487552b).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. Please note that if the perf results are neutral, you should likely undo the rollup=never given below by specifying `rollup-` to bors.

Importantly, though, if the results of this run are non-neutral **do not** roll this PR up -- it will mask other regressions or improvements in the roll up.

@bors rollup=never

Aaron1011 commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion

bors commented:
:hourglass: Trying commit 57f40e2c3a3717367024a93b7409ee32ea3c5052 with merge 8a163407b95486330c3d2cd8b9e49d07a35187c5...
<!-- homu: {"type":"TryBuildStarted","head_sha":"57f40e2c3a3717367024a93b7409ee32ea3c5052","merge_sha":"8a163407b95486330c3d2cd8b9e49d07a35187c5"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/953676136), [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=35827)
Build commit: 8a163407b95486330c3d2cd8b9e49d07a35187c5 (`8a163407b95486330c3d2cd8b9e49d07a35187c5`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=35827","checks-actions":"https://github.com/rust-lang-ci/rust/runs/953676136"},"merge_sha":"8a163407b95486330c3d2cd8b9e49d07a35187c5"} -->

rust-timer commented:
Queued 8a163407b95486330c3d2cd8b9e49d07a35187c5 with parent 3cfc7fe78eccc754b16981704a098d7bd520e2fd, future [comparison URL](https://perf.rust-lang.org/compare.html?start=3cfc7fe78eccc754b16981704a098d7bd520e2fd&end=8a163407b95486330c3d2cd8b9e49d07a35187c5).

rust-timer commented:
Finished benchmarking try commit (8a163407b95486330c3d2cd8b9e49d07a35187c5): [comparison url](https://perf.rust-lang.org/compare.html?start=3cfc7fe78eccc754b16981704a098d7bd520e2fd&end=8a163407b95486330c3d2cd8b9e49d07a35187c5).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. Please note that if the perf results are neutral, you should likely undo the rollup=never given below by specifying `rollup-` to bors.

Importantly, though, if the results of this run are non-neutral **do not** roll this PR up -- it will mask other regressions or improvements in the roll up.

@bors rollup=never

mati865 commented:
Looks like it fixes (and goes far beyond) Gimli regression for most of `-check` runs: https://perf.rust-lang.org/compare.html?start=cfc572cae2d1fc381cce476b5c787fd7221af98c&end=8a163407b95486330c3d2cd8b9e49d07a35187c5&stat=instructions%3Au

Aaron1011 commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion

bors commented:
:hourglass: Trying commit 88b56ce4834adb1de9bf2c891e7941ed550a323d with merge a5f13a6aa6cdfe3d9135bab9c67528c4cab885cd...
<!-- homu: {"type":"TryBuildStarted","head_sha":"88b56ce4834adb1de9bf2c891e7941ed550a323d","merge_sha":"a5f13a6aa6cdfe3d9135bab9c67528c4cab885cd"} -->

Aaron1011 commented:
@bors try

bors commented:
:hourglass: Trying commit 66961912f12141077444d88418c83e4dfaee4147 with merge d92cf6081e34371c22e1d61873e678d90739a6f7...
<!-- homu: {"type":"TryBuildStarted","head_sha":"66961912f12141077444d88418c83e4dfaee4147","merge_sha":"d92cf6081e34371c22e1d61873e678d90739a6f7"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/954792934)
<!-- homu: {"type":"TryBuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/954792934","builder_name":"checks-actions"} -->

Mark-Simulacrum commented:
@rust-timer build d92cf6081e34371c22e1d61873e678d90739a6f7

rust-timer commented:
Queued d92cf6081e34371c22e1d61873e678d90739a6f7 with parent 4b0882cfaa33de2796eac01afeac69097fbccbd2, future [comparison URL](https://perf.rust-lang.org/compare.html?start=4b0882cfaa33de2796eac01afeac69097fbccbd2&end=d92cf6081e34371c22e1d61873e678d90739a6f7).

rust-timer commented:
Finished benchmarking try commit (d92cf6081e34371c22e1d61873e678d90739a6f7): [comparison url](https://perf.rust-lang.org/compare.html?start=4b0882cfaa33de2796eac01afeac69097fbccbd2&end=d92cf6081e34371c22e1d61873e678d90739a6f7).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. Please note that if the perf results are neutral, you should likely undo the rollup=never given below by specifying `rollup-` to bors.

Importantly, though, if the results of this run are non-neutral **do not** roll this PR up -- it will mask other regressions or improvements in the roll up.

@bors rollup=never

Aaron1011 commented:
Perf is now either neutral or green for all non-docs builds

Aaron1011 commented:
@petrochenkov: This is now ready for review.

petrochenkov commented:
(I'll review this next weekend.)

bors commented:
:umbrella: The latest upstream changes (presumably #73851) made this pull request unmergeable. Please resolve the merge conflicts.

petrochenkov commented on src/librustc_metadata/rmeta/decoder.rs:
```diff
@@ -1723,9 +1811,6 @@ impl CrateMetadata {
         host_hash: Option<Svh>,
     ) -> CrateMetadata {
-        let def_path_table = record_time(&sess.perf_stats.decode_def_path_tables_time, || {
-            root.def_path_table.decode((&blob, sess))
```
`def_path_table` didn't actually needed the session for decoding.
So `def_key` and `def_path_hash` decoding functions don't need the session argument either.
If it's removed the patch will become noticeably cleaner.

petrochenkov commented on src/librustc_metadata/rmeta/decoder.rs:
```diff
@@ -1843,36 +1930,6 @@ impl CrateMetadata {
     }
-
-    #[inline]
-    fn def_path_hash(&self, index: DefIndex) -> DefPathHash {
```
Could you keep all the changed code in its old locations?
It's hard to read the diff if code is both changed and moved at the same time.

Aaron1011 replied:
This is not implemented on `CrateMetadata` instead of `CrateRoot`, so it needs to be moved.

petrochenkov commented on src/librustc_metadata/rmeta/decoder.rs:
```diff
@@ -1737,10 +1822,10 @@ impl CrateMetadata {
             trait_impls,
             raw_proc_macros,
             source_map_import_info: OnceCell::new(),
+            def_path_hash_map: OnceCell::new(),
```
```suggestion
            def_path_hash_map: Default::default(),
```

petrochenkov commented on src/librustc_middle/ty/context.rs:
```diff
@@ -36,7 +36,9 @@ use rustc_data_structures::sync::{self, Lock, Lrc, WorkerLocal};
-use rustc_hir::def_id::{CrateNum, DefId, DefIdMap, LocalDefId, LOCAL_CRATE};
+use rustc_hir::def_id::{
+    CrateNum, DefId, DefIdMap, LocalDefId, RawDefId, CRATE_DEF_INDEX, LOCAL_CRATE,
+};
```
Could you avoid multi-line imports?

Aaron1011 replied:
I'll change this, but I think it would be a good idea to (eventually) have `rustfmt` enforce the preferred import style.

petrochenkov commented on src/librustc_middle/ty/context.rs:
```diff
@@ -926,12 +928,11 @@ pub struct GlobalCtxt<'tcx> {
-
     pub queries: query::Queries<'tcx>,
 
+    pub local_def_path_hash_to_def_id: Option<FxHashMap<DefPathHash, DefId>>,
```
```suggestion
    pub local_def_path_hash_to_def_id: FxHashMap<DefPathHash, LocalDefId>,
```
Looks like it would be simpler to just use an empty map instead of an option.
(Also def id -> local def id.)

Aaron1011 replied:
I used an `Option` to catch any code that tries to read from this when incremental compilation is disabled.

Aaron1011 replied:
I ended up getting rid of the `Option` when I moved this to `on_disk_cache`

petrochenkov commented on src/librustc_middle/ty/context.rs:
```diff
@@ -926,12 +928,11 @@ pub struct GlobalCtxt<'tcx> {
     pub queries: query::Queries<'tcx>,
 
+    pub local_def_path_hash_to_def_id: Option<FxHashMap<DefPathHash, DefId>>,
+    def_path_hash_to_def_id_cache: Lock<FxHashMap<DefPathHash, Option<DefId>>>,
```
Could you add a comment telling that this map includes local def ids from the `local_def_path_hash_to_def_id` above? Othewise it looks like this map is for foreign def ids only.

Aaron1011 replied:
This map includes both local and foreign `DefPathHashes` - we use it to cache all lookups of a `DefPathHash`. I'll add a comment explaining tihs.

petrochenkov commented on src/librustc_middle/ty/context.rs:
```diff
@@ -1094,23 +1095,12 @@ impl<'tcx> TyCtxt<'tcx> {
+        let local_def_path_hash_to_def_id = if s.opts.build_dep_graph() {
+            let mut map = FxHashMap::with_capacity_and_hasher(
+                definitions.def_path_table().size(),
+                ::std::default::Default::default(),
```
```suggestion
                Default::default(),
```
(Not sure why the fully qualified version was used.)

petrochenkov commented on src/librustc_middle/ty/query/on_disk_cache.rs:
```diff
@@ -80,6 +80,19 @@ pub struct OnDiskCache<'sess> {
     expn_data: FxHashMap<u32, AbsoluteBytePos>,
     // Additional information used when decoding hygiene data.
     hygiene_context: HygieneDecodeContext,
+    // Maps `DefPathHash`es to their `RawDefId`s from th *previous*
```
```suggestion
    // Maps `DefPathHash`es to their `RawDefId`s from the *previous*
```

petrochenkov commented on src/librustc_middle/ty/context.rs:
```diff
@@ -1253,7 +1248,56 @@ impl<'tcx> TyCtxt<'tcx> {
+        }
+    }
+
+    pub fn get_latest_foreign_def_path_hashes(self) -> FxHashMap<DefPathHash, RawDefId> {
```
This function is unused?

petrochenkov commented on src/librustc_span/def_id.rs:
```diff
@@ -152,6 +152,19 @@ pub struct DefId {
+/// which may no longer be valid. This is used to help map a `DefPathHash`
+/// to a `DefId` in the current compilation session.
+#[derive(Encodable, Decodable, Copy, Clone, Debug)]
+pub struct RawDefId {
```
This structure is only used in `on_disk_cache.rs` so it can be moved there and privatized.

petrochenkov commented on src/librustc_metadata/rmeta/table.rs:
```diff
@@ -201,4 +201,12 @@ where
+    /// than the greatest possible index stored in this table.
+    /// This only makes sense for a dense table with no gaps
+    /// between valid indices.
+    pub fn end_index(&self) -> I {
```
```suggestion
    pub(super) fn end_index(&self) -> I {
```

petrochenkov commented on src/librustc_middle/ty/context.rs:
```diff
@@ -926,12 +928,11 @@ pub struct GlobalCtxt<'tcx> {
-
     pub queries: query::Queries<'tcx>,
 
+    pub local_def_path_hash_to_def_id: Option<FxHashMap<DefPathHash, DefId>>,
```
```suggestion
    local_def_path_hash_to_def_id: Option<FxHashMap<DefPathHash, DefId>>,
```

petrochenkov commented on src/librustc_middle/ty/query/on_disk_cache.rs:
```diff
@@ -374,6 +401,35 @@ impl<'sess> OnDiskCache<'sess> {
+        }
+    }
+
+    pub fn get_latest_foreign_def_path_hashes(&self) -> FxHashMap<DefPathHash, RawDefId> {
```
This function is also unused.

petrochenkov commented on src/librustc_middle/ty/query/on_disk_cache.rs:
```diff
@@ -374,6 +401,35 @@ impl<'sess> OnDiskCache<'sess> {
+        cnum_map[CrateNum::from_u32(cnum)]
+    }
+
+    pub fn store_foreign_def_id_hash(&self, def_id: DefId, hash: DefPathHash) {
```
```suggestion
    pub(crate) fn store_foreign_def_id_hash(&self, def_id: DefId, hash: DefPathHash) {
```

petrochenkov commented on src/librustc_middle/ty/query/on_disk_cache.rs:
```diff
@@ -374,6 +401,35 @@ impl<'sess> OnDiskCache<'sess> {
+        self.foreign_def_path_hashes.get(hash).copied()
+    }
+
+    pub fn try_remap_cnum(&self, tcx: TyCtxt<'_>, cnum: u32) -> Option<CrateNum> {
```
Perhaps some part of `def_path_hash_to_def_id` can be moved to `on_disk_cache.rs` to avoid exposing all this stuff (`fn get_raw_def_id`, `fn try_remap_cnum`) to other modules.

Aaron1011 replied:
I ended up moving all of this to `on_disk_cache.rs`

petrochenkov commented:
Did the first pass.
I still don't understand the incremental part though.

Aaron1011 commented:
@petrochenkov: I've addressed your comments. If you have any specific questions about the incremental part, I can add additional comments to the relevenat code.

petrochenkov commented:
This PR grew some pure refactoring parts now, it would be great if you could move them to a separate PR.
I'll do the second pass through the main part of this PR, including incremental, some time closer to the weekend.

Aaron1011 commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion

bors commented:
:hourglass: Trying commit 10f0e369dfb19ac8271f049788a27e43d94cc9ed with merge 2567aeba3d5cc52337c6826ab7e5e9ed6846fbc3...
<!-- homu: {"type":"TryBuildStarted","head_sha":"10f0e369dfb19ac8271f049788a27e43d94cc9ed","merge_sha":"2567aeba3d5cc52337c6826ab7e5e9ed6846fbc3"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1023124569), [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=36189)
Build commit: 2567aeba3d5cc52337c6826ab7e5e9ed6846fbc3 (`2567aeba3d5cc52337c6826ab7e5e9ed6846fbc3`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=36189","checks-actions":"https://github.com/rust-lang-ci/rust/runs/1023124569"},"merge_sha":"2567aeba3d5cc52337c6826ab7e5e9ed6846fbc3"} -->

rust-timer commented:
Queued 2567aeba3d5cc52337c6826ab7e5e9ed6846fbc3 with parent 03017003c77d782cf7ed841a05d7c628a9b93f25, future [comparison URL](https://perf.rust-lang.org/compare.html?start=03017003c77d782cf7ed841a05d7c628a9b93f25&end=2567aeba3d5cc52337c6826ab7e5e9ed6846fbc3).

petrochenkov commented:
PR title and description need updating.

rust-timer commented:
Finished benchmarking try commit (2567aeba3d5cc52337c6826ab7e5e9ed6846fbc3): [comparison url](https://perf.rust-lang.org/compare.html?start=03017003c77d782cf7ed841a05d7c628a9b93f25&end=2567aeba3d5cc52337c6826ab7e5e9ed6846fbc3).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. Please note that if the perf results are neutral, you should likely undo the rollup=never given below by specifying `rollup-` to bors.

Importantly, though, if the results of this run are non-neutral **do not** roll this PR up -- it will mask other regressions or improvements in the roll up.

@bors rollup=never

petrochenkov commented:
Redirecting this to @rust-lang/wg-incr-comp.
r? @pnkfelix 

I've also seen some traces of @michaelwoerister recently :)
It would be really great if you could look at this.

Aaron1011 commented:
@petrochenkov: I've updated the title and description

wesleywiser commented:
Thanks @petrochenkov! We have a meeting scheduled for tonight so I've added it to the agenda to discuss.

bors commented:
:umbrella: The latest upstream changes (presumably #76233) made this pull request unmergeable. Please resolve the merge conflicts.

crlf0710 commented:
Triage: There's merge conflicts now.

crlf0710 commented:
Triage: CI is still red here.

Dylan-DPC commented:
@Aaron1011 any updates?

Aaron1011 commented:
Rebased

Aaron1011 commented:
@wesleywiser Did this PR end up getting discussed?

wesleywiser commented:
@Aaron1011 I took a look at it yesterday and the code seems fine to me but I don't think I have a complete understanding of this part of the incremental system so I shouldn't sign off on it. @pnkfelix said he would do the review and sign off.

Aaron1011 commented:
It's been a while since the last perf run. Let's make sure nothing's changed:

@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion

bors commented:
:hourglass: Trying commit acf68e1096d8b2a4d629561e045520790d67dadc with merge c2a99f87ee370b7466a1ba8fef305771c7bebe8e...
<!-- homu: {"type":"TryBuildStarted","head_sha":"acf68e1096d8b2a4d629561e045520790d67dadc","merge_sha":"c2a99f87ee370b7466a1ba8fef305771c7bebe8e"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1302674148)
Build commit: c2a99f87ee370b7466a1ba8fef305771c7bebe8e (`c2a99f87ee370b7466a1ba8fef305771c7bebe8e`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/1302674148"},"merge_sha":"c2a99f87ee370b7466a1ba8fef305771c7bebe8e"} -->

rust-timer commented:
Queued c2a99f87ee370b7466a1ba8fef305771c7bebe8e with parent 2e8a54af60df63034e41359acfc923e5c5769a91, future [comparison URL](https://perf.rust-lang.org/compare.html?start=2e8a54af60df63034e41359acfc923e5c5769a91&end=c2a99f87ee370b7466a1ba8fef305771c7bebe8e).

rust-timer commented:
Finished benchmarking try commit (c2a99f87ee370b7466a1ba8fef305771c7bebe8e): [comparison url](https://perf.rust-lang.org/compare.html?start=2e8a54af60df63034e41359acfc923e5c5769a91&end=c2a99f87ee370b7466a1ba8fef305771c7bebe8e).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. Please note that if the perf results are neutral, you should likely undo the rollup=never given below by specifying `rollup-` to bors.

Importantly, though, if the results of this run are non-neutral **do not** roll this PR up -- it will mask other regressions or improvements in the roll up.

@bors rollup=never

bjorn3 commented:
Huge wins of up to 35% in incremental mode. Slight regression of up to 0.5% in non-incremental mode.

Aaron1011 commented:
It looks like some of this code doesn't actually need to run in non-incremental mode. I'll work on disabling it.

pnkfelix commented on compiler/rustc_middle/src/ty/query/on_disk_cache.rs:
```diff
@@ -374,6 +444,31 @@ impl<'sess> OnDiskCache<'sess> {
+
+    pub(crate) fn store_foreign_def_id_hash(&self, def_id: DefId, hash: DefPathHash) {
+        // We may overwrite an existing entry, but it will have the same value,
+        // so it's fine
```
should we debug_assert the above statement?

pnkfelix replied:
(then again, maybe not; I forgot that this is being called as a side-effect of `to_fingerprint`, and I was already concerned about whether we were adding excessive overhead there...)

pnkfelix commented:
We discussed this at the most recent wg-incr-comp meeting.

We are generally inclined to land this (the performance benefits are after all **huge**), but we also want to do it early in a release cycle; i.e., shortly after the next beta is cut.

In the meantime: @Aaron1011 , have you had a chance to look into the revisions you [alluded to above](https://github.com/rust-lang/rust/pull/74967#issuecomment-716059416)?

Aaron1011 commented:
Blocked on https://github.com/rust-lang/rust/pull/79216

Aaron1011 commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion

bors commented:
:hourglass: Trying commit 7a9aa4f9808fe37afb581ba54884612df2034a75 with merge bf8e95436e60effbeb46a32e17df8ab7fcb0c6ad...
<!-- homu: {"type":"TryBuildStarted","head_sha":"7a9aa4f9808fe37afb581ba54884612df2034a75","merge_sha":"bf8e95436e60effbeb46a32e17df8ab7fcb0c6ad"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1455777182)
Build commit: bf8e95436e60effbeb46a32e17df8ab7fcb0c6ad (`bf8e95436e60effbeb46a32e17df8ab7fcb0c6ad`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/1455777182"},"merge_sha":"bf8e95436e60effbeb46a32e17df8ab7fcb0c6ad"} -->

rust-timer commented:
Queued bf8e95436e60effbeb46a32e17df8ab7fcb0c6ad with parent db79d2f63780613e700cb58b4339c48287555ae0, future [comparison URL](https://perf.rust-lang.org/compare.html?start=db79d2f63780613e700cb58b4339c48287555ae0&end=bf8e95436e60effbeb46a32e17df8ab7fcb0c6ad).

rust-timer commented:
Finished benchmarking try commit (bf8e95436e60effbeb46a32e17df8ab7fcb0c6ad): [comparison url](https://perf.rust-lang.org/compare.html?start=db79d2f63780613e700cb58b4339c48287555ae0&end=bf8e95436e60effbeb46a32e17df8ab7fcb0c6ad).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. Please note that if the perf results are neutral, you should likely undo the rollup=never given below by specifying `rollup-` to bors.

Importantly, though, if the results of this run are non-neutral **do not** roll this PR up -- it will mask other regressions or improvements in the roll up.

@bors rollup=never
@rustbot modify labels: +S-waiting-on-review -S-waiting-on-perf

Aaron1011 commented:
Regressions of up to 0.1% in non-incrememtal mode.

This is now ready for review again.

pnkfelix commented on compiler/rustc_metadata/src/rmeta/decoder.rs:
```diff
@@ -77,6 +77,10 @@ crate struct CrateMetadata {
     /// Source maps for code from the crate.
     source_map_import_info: OnceCell<Vec<ImportedSourceFile>>,
+    /// For every definition in this crate, maps its `DefPathHash` to its
+    /// `DefIndex`. See `raw_def_id_to_def_id` for more details about how
```
should this comment say `def_path_hash_to_def_id` instead?

pnkfelix commented on compiler/rustc_query_system/src/dep_graph/graph.rs:
```diff
@@ -699,6 +700,18 @@ impl<K: DepKind> DepGraph<K> {
+        // We have just loaded a deserialized `DepNode` from the previous
+        // compilation session into the current one. If this was a foreign `DefId`,
+        // then we stored additional information in the incr comp cache when we
+        // initially created its fingerprint (see `DepNodeParams::to_fingerprint`)
```
super super nit: missing a period ending the sentence after the parenthesis here, making this whole paragraph look like one big run-on sentence.

pnkfelix commented:
@bors r+

bors commented:
:pushpin: Commit 7a9aa4f9808fe37afb581ba54884612df2034a75 has been approved by `pnkfelix`

<!-- @bors r=pnkfelix 7a9aa4f9808fe37afb581ba54884612df2034a75 -->
<!-- homu: {"type":"Approved","sha":"7a9aa4f9808fe37afb581ba54884612df2034a75","approver":"pnkfelix"} -->

bors commented:
:hourglass: Testing commit 7a9aa4f9808fe37afb581ba54884612df2034a75 with merge 4cbda829c00af2c3ac362c979fa97ea90be0be7d...
<!-- homu: {"type":"BuildStarted","head_sha":"7a9aa4f9808fe37afb581ba54884612df2034a75","merge_sha":"4cbda829c00af2c3ac362c979fa97ea90be0be7d"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1481289400)
Approved by: pnkfelix
Pushing 4cbda829c00af2c3ac362c979fa97ea90be0be7d to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"pnkfelix","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/1481289400"},"merge_sha":"4cbda829c00af2c3ac362c979fa97ea90be0be7d"} -->
